### PR TITLE
Sequence Parallel for LoRA

### PR DESCRIPTION
**Pull request by cuichenx**
# What does this PR do ?

Allow sequence parallel for lora training. Supports all linear modules as in #7988

**Collection**: NLP

# Changelog 
- Set sequence_parallel flags to False in LoRA layers to allow gathering output in LoRA `linear_in`
- For attention_qkv and mlp_fc1, do a manual all_gather (reduce_scatter during backward) along the sequence dimension before each lora module since the previous layer (layernorm) is affected by SP
- For attention_dense and mlp_fc2, do a manually scatter (gather during backward) along the sequence dimension after each lora module since the next layer (layernorm) is affected by SP

# Jenkins CI
To run Jenkins, a NeMo User with write access must comment `jenkins` on the PR.

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # 7988

### CODE REVIEW

**Comment by cuichenx** [comment]
jenkins

**Comment by cuichenx** [comment]
jenkins

**Comment by cuichenx** [comment]
jenkins

**Comment by cuichenx** [comment]
jenkins

**Comment by cuichenx** [comment]
Known issue: `model.ub_tp_comm_overlap=True` is not supported for LoRA + SP

**Comment by cuichenx** [comment]
jenkins

**Comment by cuichenx** [comment]
jenkins

**Review by arendu — APPROVED** [review]
LGTM!

File: nemo/collections/nlp/modules/common/megatron/adapters/parallel_adapters.py [hunk 1]
```diff
@@ -224,6 +230,11 @@ def forward(self, x):
 
         if self.norm_position == 'pre':
             x = self.layer_norm(x)
+        if self._sequence_parallel:
+            # layernorm before lora is impacted by sequence parallel,
+            # hence seq dim need to be gathered right before lora linear layers
+            # this function also handles the backward pass correctly
+            x = gather_from_sequence_parallel_region(x)
```
[the remark this answers is not in the document]
Comment (cuichenx) [reply nemo/collections/nlp/modules/common/megatron/adapters/parallel_adapters.py:257]: 1. It's doing repeated gathers but the input is an ungathered tensor. This is bad for performance. I'm looking into addressing this with the perf team.
2. `self._sequence_parallel` will reflect the training config. It's always SP split for `attention_qkv` and `mlp_fc1` and never for `attention_dense` or `mlp_fc2`. This corresponds with `self.input_is_parallel` so I've added a check there

**Pull request merged**